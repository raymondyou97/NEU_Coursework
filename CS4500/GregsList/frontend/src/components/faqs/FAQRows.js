import PropTypes from 'prop-types';
import React, { useRef, useState } from 'react';
import { Button } from 'react-bootstrap';

const FAQRows = ({ faqs, onUpdate, onDelete }) => {
  const [editId, setEditId] = useState(undefined);
  const titleInputRef = useRef(undefined);
  const questionInputRef = useRef(undefined);

  return faqs.map(faq => {
    const editing = editId === faq.id;
    return (
      <tr key={faq.id}>
        <td>
          {editing ? (
            <input ref={titleInputRef} defaultValue={faq.title} />
          ) : (
            faq.title
          )}
        </td>
        <td>
          {editing ? (
            <input ref={questionInputRef} defaultValue={faq.question} />
          ) : (
            faq.question
          )}
        </td>
        <td>
          <Button
            onClick={
              editing
                ? // Save the edits to this row
                  () => {
                    onUpdate(faq.id, {
                      title: titleInputRef.current.value,
                      question: questionInputRef.current.value,
                    });
                    setEditId(undefined);
                  }
                : // Start editing this row
                  () => setEditId(faq.id)
            }
            size="sm"
            style={{ height: 'auto' }}
          >
            {editing ? 'Save' : 'Edit'}
          </Button>
        </td>
        <td>
          <Button
            onClick={() => onDelete(faq.id)}
            variant="danger"
            size="sm"
            style={{ height: 'auto' }}
          >
            X
          </Button>
        </td>
      </tr>
    );
  });
};

FAQRows.propTypes = {
  faqs: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      title: PropTypes.string.isRequired,
      question: PropTypes.string.isRequired,
    }).isRequired
  ).isRequired,
  onUpdate: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
};

export default FAQRows;
