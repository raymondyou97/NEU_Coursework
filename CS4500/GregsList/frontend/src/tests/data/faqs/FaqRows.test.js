import React from 'react';
import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';

import FAQRows from '../../../components/faqs/FAQRows';
import { makeFaqs } from './FaqData';

test('FAQRows', () => {
  const faqs = makeFaqs();

  const component = shallow(<FAQRows faqs={faqs} />);
  expect(component.find('tr').length).toEqual(faqs.length);

  const firstRow = component.find('tr').first();
  expect(
    firstRow
      .find('td')
      .at(0)
      .text()
  ).toEqual(faqs[0].title);
  expect(
    firstRow
      .find('td')
      .at(1)
      .text()
  ).toEqual(faqs[0].question);
});

test('Snapshot renders', () => {
  const faqs = makeFaqs();

  const component = renderer.create(<FAQRows faqs={faqs} />);

  let componentTree = component.toJSON();
  expect(componentTree).toMatchSnapshot();
});
