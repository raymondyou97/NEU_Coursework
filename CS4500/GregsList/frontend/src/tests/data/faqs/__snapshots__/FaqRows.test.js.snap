// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot renders 1`] = `
Array [
  <tr>
    <td>
      Question #0
    </td>
    <td>
      What is your favorite color?
    </td>
    <td>
      <button
        className="btn btn-primary btn-sm"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "height": "auto",
          }
        }
        type="button"
      >
        Edit
      </button>
    </td>
    <td>
      <button
        className="btn btn-danger btn-sm"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "height": "auto",
          }
        }
        type="button"
      >
        X
      </button>
    </td>
  </tr>,
  <tr>
    <td>
      Question #1
    </td>
    <td>
      What is your favorite color?
    </td>
    <td>
      <button
        className="btn btn-primary btn-sm"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "height": "auto",
          }
        }
        type="button"
      >
        Edit
      </button>
    </td>
    <td>
      <button
        className="btn btn-danger btn-sm"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "height": "auto",
          }
        }
        type="button"
      >
        X
      </button>
    </td>
  </tr>,
  <tr>
    <td>
      Question #2
    </td>
    <td>
      What is your favorite color?
    </td>
    <td>
      <button
        className="btn btn-primary btn-sm"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "height": "auto",
          }
        }
        type="button"
      >
        Edit
      </button>
    </td>
    <td>
      <button
        className="btn btn-danger btn-sm"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "height": "auto",
          }
        }
        type="button"
      >
        X
      </button>
    </td>
  </tr>,
  <tr>
    <td>
      Question #3
    </td>
    <td>
      What is your favorite color?
    </td>
    <td>
      <button
        className="btn btn-primary btn-sm"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "height": "auto",
          }
        }
        type="button"
      >
        Edit
      </button>
    </td>
    <td>
      <button
        className="btn btn-danger btn-sm"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "height": "auto",
          }
        }
        type="button"
      >
        X
      </button>
    </td>
  </tr>,
  <tr>
    <td>
      Question #4
    </td>
    <td>
      What is your favorite color?
    </td>
    <td>
      <button
        className="btn btn-primary btn-sm"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "height": "auto",
          }
        }
        type="button"
      >
        Edit
      </button>
    </td>
    <td>
      <button
        className="btn btn-danger btn-sm"
        disabled={false}
        onClick={[Function]}
        style={
          Object {
            "height": "auto",
          }
        }
        type="button"
      >
        X
      </button>
    </td>
  </tr>,
]
`;
